### PR TITLE
Implement memory pool index compaction

### PR DESCRIPTION
**Pull request by okuta**
The size of `SingleDeviceMemoryPool._free` increases in some case. Sometimes it may exceed 100000.
Therefore malloc gets slower when we use CuPy long time. Especially in Chainer's workload there is a huge performance degradation.

In this PR, I suggest using None instead of empty `set` object and compressing `_free` arrays.

### CODE REVIEW

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -514,6 +516,40 @@ class PooledMemory(Memory):
     __del__ = free
 
 
+cdef int _compaction_threshold = 512
+
+
+cdef _compaction(SingleDeviceMemoryPool pool, size_t stream_ptr, bint free):
```
Comment (sonots) [inline cupy/cuda/memory.pyx]: The term memory compaction reminds me something like https://docs.oracle.com/cd/E13150_01/jrockit_jvm/jrockit/geninfo/diagnos/garbage_collect.html#wp1086917.
Also, we may implement in the future https://github.com/cupy/cupy/issues/536.

Using another term would be better? How about "index_compaction"? or "compact_arena"?

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -514,6 +516,40 @@ class PooledMemory(Memory):
...
+                continue
+            free_list = keep_list
+
+        new_index.push_back(arena_index.at(index))
+        new_arena.append(free_list)
+    arena_index.swap(new_index)
+    arena[:] = new_arena
+
+
+
```
Comment (sonots) [inline cupy/cuda/memory.pyx]: flake8 fails

**Comment by sonots** [comment]
Basically LGTM

**Comment by okuta** [comment]
Thanks.
I fixed.

**Comment by sonots** [comment]
LGTM

**Pull request merged**